src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx

```diff
@@ -0,0 +1,150 @@
+import * as React from 'react';
+import { Button, TextContent, Text, Grid, GridItem, Bullseye, Flex } from '@patternfly/react-core';
+import { PlusCircleIcon, TrashIcon } from '@patternfly/react-icons';
+import spacing from '@patternfly/react-styles/css/utilities/Spacing/spacing';
+import { MappingType, MappingSource, MappingTarget } from '../../types';
+import { IVMwareNetwork, IVMwareDatastore } from '@app/Providers/types';
+import LineArrow from '@app/common/components/LineArrow/LineArrow';
+import MappingSourceSelect from './MappingSourceSelect';
+import MappingTargetSelect from './MappingTargetSelect';
+import './MappingBuilder.css';
+
+export interface IMappingBuilderItem {
+  source: MappingSource | null;
+  target: MappingTarget | null;
+}
+
+interface IMappingBuilderProps {
+  mappingType: MappingType;
+  availableSources: MappingSource[];
+  availableTargets: MappingTarget[];
+  builderItems: IMappingBuilderItem[];
+  setBuilderItems: (groups: IMappingBuilderItem[]) => void;
+}
+
+export const MappingBuilder: React.FunctionComponent<IMappingBuilderProps> = ({
+  mappingType,
+  availableSources,
+  availableTargets,
+  builderItems,
+  setBuilderItems,
+}: IMappingBuilderProps) => {
+  const reset = () => setBuilderItems([{ source: null, target: null }]);
+  const isReset = builderItems.length === 1 && !builderItems[0].source && !builderItems[0].target;
+  const addEmptyItem = () => setBuilderItems([...builderItems, { source: null, target: null }]);
+  const removeItem = (itemIndex: number) => {
+    if (builderItems.length > 1) {
+      setBuilderItems(builderItems.filter((_item, index) => index !== itemIndex));
+    } else {
+      reset();
+    }
+  };
+
+  let instructionText = '';
+  let sourceHeadingText = '';
+  let targetHeadingText = '';
+  let selectSourcePlaceholder = '';
+  let selectTargetPlaceholder = '';
+  if (mappingType === MappingType.Network) {
+    instructionText = 'Map source and target networks.';
+    sourceHeadingText = 'Source networks';
+    targetHeadingText = 'Target networks';
+    selectSourcePlaceholder = 'Select source...';
+    selectTargetPlaceholder = 'Select target...';
+  }
+  if (mappingType === MappingType.Storage) {
+    instructionText = 'Map source datastores to target storage classes.';
+    sourceHeadingText = 'Source datastores';
+    targetHeadingText = 'Target storage classes';
+    selectSourcePlaceholder = 'Select source...';
+    selectTargetPlaceholder = 'Select target...';
+  }
+
+  return (
+    <>
+      <TextContent>
+        <Text component="p">{instructionText}</Text>
+      </TextContent>
+      {builderItems.map((item, itemIndex) => {
+        let key = '';
+        if (mappingType === MappingType.Network) {
+          const t = item.source as IVMwareNetwork | null;
```

Comment:
:+1: Very cool use of typescript type assertion.

Reply:
Honestly though I can't remember why I used the variable name `t`. I think it came from using it on the target side, but this one should be `s` or I should use longer names. I'll add a commit.

Reply:
Actually, turns out I didn't need this at all. Both types of sources have an `id` property. I'll remove the whole condition